core/src/main/java/org/opensearch/sql/planner/physical/TrendlineOperator.java

```diff
@@ -0,0 +1,324 @@
+  @EqualsAndHashCode.Exclude private final HashSet<String> aliases;
+
+  public TrendlineOperator(
+      PhysicalPlan input, List<Pair<Trendline.TrendlineComputation, ExprCoreType>> computations) {
+    this.input = input;
+    this.computations = computations;
+    this.accumulators = computations.stream().map(TrendlineOperator::createAccumulator).toList();
+    fieldToIndexMap = new HashMap<>(computations.size());
+    aliases = new HashSet<>(computations.size());
+    for (int i = 0; i < computations.size(); ++i) {
+      final Trendline.TrendlineComputation computation = computations.get(i).getKey();
+      fieldToIndexMap.put(computation.getDataField().getChild().get(0).toString(), i);
+      aliases.add(computation.getAlias());
+    }
+  }
+
+  @Override
+  public <R, C> R accept(PhysicalPlanNodeVisitor<R, C> visitor, C context) {
+    return visitor.visitTrendline(this, context);
+  }
+
+  @Override
+  public List<PhysicalPlan> getChild() {
+    return Collections.singletonList(input);
+  }
+
+  @Override
+  public boolean hasNext() {
+    return getChild().getFirst().hasNext();
+  }
+
+  @Override
+  public ExprValue next() {
+    final ExprValue result;
+    final ExprValue next = input.next();
+    final Map<String, ExprValue> inputStruct = consumeInputTuple(next);
+    final Builder<String, ExprValue> mapBuilder = new Builder<>();
+    mapBuilder.putAll(inputStruct);
+
+    // Add calculated trendline values, which might overwrite existing fields from the input.
+    for (int i = 0; i < accumulators.size(); ++i) {
+      final ExprValue calculateResult = accumulators.get(i).calculate();
+      final String field = computations.get(i).getKey().getAlias();
+      if (calculateResult != null) {
+        mapBuilder.put(field, calculateResult);
+      }
+    }
+
+    result = ExprTupleValue.fromExprValueMap(mapBuilder.buildKeepingLast());
+    return result;
+  }
+
+  private Map<String, ExprValue> consumeInputTuple(ExprValue inputValue) {
+    final Map<String, ExprValue> tupleValue = ExprValueUtils.getTupleValue(inputValue);
+    for (String bindName : tupleValue.keySet()) {
+      final Integer index = fieldToIndexMap.get(bindName);
+      if (index != null) {
+        accumulators.get(index).accumulate(tupleValue.get(bindName));
+      }
+    }
+    tupleValue.keySet().removeAll(aliases);
+    return tupleValue;
+  }
+
+  private static TrendlineAccumulator createAccumulator(
+      Pair<Trendline.TrendlineComputation, ExprCoreType> computation) {
+    switch (computation.getKey().getComputationType()) {
+      case SMA:
+        return new SimpleMovingAverageAccumulator(computation.getKey(), computation.getValue());
+      case WMA:
+      default:
+        throw new IllegalStateException(
+            "Unexpected value: " + computation.getKey().getComputationType());
+    }
+  }
+
+  /** Maintains stateful information for calculating the trendline. */
+  private interface TrendlineAccumulator {
```

Comment:
I'm leaving it here since these APIs are tightly-coupled to Trendline and not likely to be used independently. I don't have a strong opinion on this though.